### PR TITLE
gh-111389: Add PyHASH_MULTIPLIER constant

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-111389

----
📚 Documentation preview 📚: https://cpython-previews--119214.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
The private `_PyHASH_MULTIPLIER` constant was removed in Python 3.13 alpha 1, but [had to be reverted](https://github.com/python/cpython/pull/112115) since it broke too many projects.

A code search on PyPI top 7,500 projects (2024-03-16) finds 4 projects using it:

* asyncpg (0.29.0)
* numba (0.59.0)
* orjson (3.9.15)
* recordclass (0.21.1)

File: Include/cpython/pyhash.h [hunk 1]
```diff
@@ -3,7 +3,7 @@
 #endif
 
 /* Prime multiplier used in string and various other hashes. */
-#define _PyHASH_MULTIPLIER 1000003UL  /* 0xf4243 */
+#define PyHASH_MULTIPLIER 1000003UL  /* 0xf4243 */
```
Comment (gpshead) [inline Include/cpython/pyhash.h:6]: Should we keep the legacy _ name around (but undocumented) to smooth the transition for those few projects that use it in their existing extension code?

`#define _PyHASH_MULTIPLIER PyHASH_MULTIPLIER`

Comment (gpshead) [reply Include/cpython/pyhash.h:6]: oh wait, you already did this below. nevermind! :)

**Review by gpshead — APPROVED** [review]
This makes sense, no harm in this being public even if I think it is _odd_ that anyone uses it.

**Comment by vstinner** [comment]
> This makes sense, no harm in this being public even if I think it is odd that anyone uses it.

It's a good prime number to increase to implement good hash function. Using a random number can lead to more collision.

cc @mdickinson

**Comment by mdickinson** [comment]
> cc @mdickinson

Change seems reasonable to me; the value is potentially useful to have available if you're implementing something complex-like and need equality to behave well with respect to Python's `complex`. (I'll confess that I've never seen this need in practice, but I've seen the equivalent for float hashing.)

**Comment by vstinner** [comment]
>  (I'll confess that I've never seen this need in practice

I don't know about the specific case of floating point objects, but the constant is used by at least 4 projects: https://github.com/python/cpython/pull/119214#issuecomment-2120609403

**Pull request merged**